src/core/src/bound_evaluate.cpp

```diff
@@ -75,85 +75,219 @@ bool is_type_allocable(const element::Type& type) {
-                        tensor.invalidate_values();
-                }
-                propagate_rt_info(node, output);
+        for (const auto& output : node.outputs()) {
+            if (is_type_allocable(output.get_element_type())) {
+                lowers.emplace_back(output);
+                uppers.emplace_back(output);
             } else {
-                break;
+                lowers.emplace_back();
+                uppers.emplace_back();
+            }
+        }
+    }
+
+    void set_bound_at_port(descriptor::Tensor& tensor_desc, const size_t port, const bool same_inputs) const {
+        tensor_desc.set_lower_value(lowers[port]);
+        if (same_inputs || are_equal(lowers[port], uppers[port])) {
+            tensor_desc.set_upper_value(lowers[port]);
+        } else {
+            tensor_desc.set_upper_value(uppers[port]);
+        }
+    }
+
+    bool evaluate(const Node& node) {
+        return node.evaluate_lower(lowers) && node.evaluate_upper(uppers);
+    }
+
+    ov::TensorVector lowers, uppers;
+};
+
+/** @brief BoundEvaluator evaluate and set bound depends specified BoundType. */
+template <class BoundType>
+struct BoundEvaluator : BoundType {
+    void init(const Node* n) {
+        node = n;
+        symbols.resize(0);
+        BoundType::init(*node);
+    }
+
+    void set_bounds_and_symbols() const {
+        const auto same_inputs = node_has_same_inputs(*node);
+        const auto symbols_evaluated = !symbols.empty();
+
+        for (size_t port = 0; port < node->get_output_size(); ++port) {
+            auto& output_tensor_desc = node->get_output_tensor(port);
+
+            BoundType::set_bound_at_port(output_tensor_desc, port, same_inputs);
+
+            if (symbols_evaluated && output_tensor_desc.get_value_symbol().empty()) {
+                output_tensor_desc.set_value_symbol(symbols[port]);
+                ov::populate_tensor_with_missing_symbols(output_tensor_desc);
+            }
+        }
+    }
+
+    bool evaluate() {
+        if (BoundType::evaluate(*node)) {
+            symbols.resize(node->get_output_size());
+            if (!node->evaluate_symbol(symbols)) {
+                symbols.resize(0);
             }
+            return true;
+        } else {
+            return false;
         }
     }
 
-    if (is_upper)
-        return output.get_tensor().get_upper_value();
-    else
-        return output.get_tensor().get_lower_value();
+    const Node* node;
+    TensorSymbolVector symbols;
+};
+
+/**
+ * @brief Evaluate bound(s) algorithm.
+ *
+ * Calculate required bounds for Node's output.
+ *
+ * @tparam        BoundEvaluator Specify the bound type evaluator.
+ * @param output  Node's output for which bounds will be calculated.
+ */
+template <class BoundEvaluator>
+void evaluate_bound(const Output<Node>& output) {
+    std::vector<Node*> ordered_nodes;
+    if (BoundEvaluator::requires_evaluation(output.get_tensor()) && could_propagate(output, ordered_nodes)) {
```

Comment:
To make the propagation even more efficient -- we can add a check for `has_evaluate` into the `could_propagate` routine.

Reply:
The `could_propagate` will not add nodes with bound to ordered nodes